### PR TITLE
[SYCL][L0] Programming guide for multi-tile and multi-card under Level-Zero backend.

### PR DESCRIPTION
**Pull request by smaslov-intel**
Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/doc/MultiTileCardWithLevelZero.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+[level_zero:0] GPU : Intel(R) Level-Zero 1.1 [1.1.19792]
+```
+Another way is to use similar SYCL API described in [FilterSelector.md](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/FilterSelector/FilterSelector.adoc)
+E.g. `filter_selector("level_zero")` will only see Level-Zero operated devices.
+
+If there are multiple GPUs in a system then they will be seen as multiple different root-devices.
+On Linux these would be multiple SYCL root-devices of the same SYCL platform (representing Level-Zero driver).
+On Windows these would appear as root-devices of multiple different SYCL platforms (Level-Zero drivers).
+
+`CreateMultipleRootDevices=N` evironment variable can be used to emulate multiple GPU cards.
```
Comment (vladimirlaz) [inline sycl/doc/MultiTileCardWithLevelZero.md:34]: Could you add comment/note that NEOReadDebugKeys=1 variable should be also set to make RT read mentioned debug options?

**Review by alexbatashev — CHANGES_REQUESTED** [review]
@smaslov-intel could you please update index.rst file in the same directory? It is required for correct rendering to https://intel.github.io/llvm-docs/

File: sycl/doc/MultiTileCardWithLevelZero.md (hunk 1 above)
Comment (smaslov-intel) [reply sycl/doc/MultiTileCardWithLevelZero.md]: Added.

**Comment by smaslov-intel** [comment]
> @smaslov-intel could you please update index.rst file in the same directory? It is required for correct rendering to https://intel.github.io/llvm-docs/

Added.

File: sycl/doc/MultiTileCardWithLevelZero.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+[level_zero:0] GPU : Intel(R) Level-Zero 1.1 [1.1.19792]
+```
+Another way is to use similar SYCL API described in [FilterSelector.md](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/FilterSelector/FilterSelector.adoc)
+E.g. `filter_selector("level_zero")` will only see Level-Zero operated devices.
+
+If there are multiple GPUs in a system then they will be seen as multiple different root-devices.
+On Linux these would be multiple SYCL root-devices of the same SYCL platform (representing Level-Zero driver).
+On Windows these would appear as root-devices of multiple different SYCL platforms (Level-Zero drivers).
+
+`CreateMultipleRootDevices=N NEOReadDebugKeys=1` evironment variables can be used to emulate multiple GPU cards.
```
Comment (pvchupin) [inline sycl/doc/MultiTileCardWithLevelZero.md:34]: Can we add example of command/output here like for sycl-ls above?

File: sycl/doc/MultiTileCardWithLevelZero.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+}
+```
+
+Each call to `create_sub_devices` will return exactly the same sub-devices and in the persistent order.
+To control what sub-devices are exposed by Level-Zero UMD one can use ZE_AFFINITY_MASK environment variable.
+
+NOTE: The `partition_by_affinity_domain` is the only partitioning supported for Intel GPU.
+Similar `next_partitionable` and `numa` are the only partitioning properties supported (both doing the same thing).
+
+`CreateMultipleSubDevices=N NEOReadDebugKeys=1` environment variables can be used to emulate multiple tiles of a GPU.
```
Comment (pvchupin) [inline sycl/doc/MultiTileCardWithLevelZero.md:55]: same here, some example might be helpful.

File: sycl/doc/MultiTileCardWithLevelZero.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+- Memory buffers for context with sub-devices of the same root-device (possibly including the root-device itself) are allocated on that root-device.
+   Thus they are readily accessible by all the devices in such context. The synchronization with the host is performed by SYCL RT with map/unmap doing implicit copies when necessary.
+- Memory buffers for context with devices from different root-devices in it are allocated on host (thus made accessible to all devices).
+	
+## 4 Queue
+
+SYCL queue is always attached to a single device in a possibly multi-device context.
+Some typical scenarios are the following (from most performant to least performant):
+
+A. Context with a single sub-device in it and the queue is attached to that sub-device (tile)
```
Comment (pvchupin) [inline sycl/doc/MultiTileCardWithLevelZero.md:103]: Same here. Might be a good idea to inline short snippets of code with examples.

File: sycl/doc/MultiTileCardWithLevelZero.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+		
+D. Contexts with multiple root-devices (multi-card)
+- The most unrestrictive context with queues attached to different root-devices
+- Offers most sharing possibilities at the cost of slow access through host memory or explicit copies needed
+
+Depending on the chosen programming model (A,B,C,D) and algorithm used make sure to do proper memory allocation/synchronization.
+				
+## 5 Examples
+	
+These are few examples of programming to multiple tiles and multiple cards:
```
Comment (pvchupin) [inline sycl/doc/MultiTileCardWithLevelZero.md:123]: These are external examples which are good, but might be lost over time... can we have something as part of this doc?

File: sycl/doc/MultiTileCardWithLevelZero.md (hunk 2 above)
Comment (smaslov-intel) [reply sycl/doc/MultiTileCardWithLevelZero.md:34]: OK.

File: sycl/doc/MultiTileCardWithLevelZero.md (hunk 3 above)
Comment (smaslov-intel) [reply sycl/doc/MultiTileCardWithLevelZero.md:55]: Today "sycl-ls" doesn't print info about sub-devices, so I can't give short example

File: sycl/doc/MultiTileCardWithLevelZero.md (hunk 5 above)
Comment (smaslov-intel) [reply sycl/doc/MultiTileCardWithLevelZero.md:123]: I wanted to shoe E2E test these would be too large for this doc

File: sycl/doc/MultiTileCardWithLevelZero.md (hunk 4 above)
Comment (smaslov-intel) [reply sycl/doc/MultiTileCardWithLevelZero.md:108]: OK, added

**Comment by smaslov-intel** [comment]
Have no idea why CI is broken:
> fatal: unable to access 'https://github.com/KhronosGroup/OpenCL-Headers/': Couldn't connect to server

**Comment by pvchupin** [comment]
@tfzhu, one more issue

**Comment by pvchupin** [comment]
> @tfzhu, one more issue

Same as for other PRs restart helped.

**Review by pvchupin — APPROVED** [review]
Would be good to get more reviews from others.

File: sycl/doc/MultiTileCardWithLevelZero.md (hunk 4 above)
Comment (smaslov-intel) [reply sycl/doc/MultiTileCardWithLevelZero.md]: added

**Review by vladimirlaz — APPROVED** [review]

**Review by alexbatashev — APPROVED** [review]

**Pull request merged**